### PR TITLE
HYDRA-42 Commit output only once either in destroy or in onFinish method.

### PR DESCRIPTION
**Pull request by sagarkapare**
We were committing output twice once in `destory` and then in `onFinish` method which was causing the ERRORS mentioned in JIRA - https://issues.cask.co/browse/HYDRA-42. There errors are harmless though.

### CODE REVIEW

**Comment by anew** [comment]
I don't know whether I agree with this change. The duplicate commit was introduced by adding the destroy() and copying the code over. This was not right. Instead, what we should do is implement a 
finish() method that:
1. commits all datasets that implement OutputCommitterProvider
2. if the mapreduce implements ProgramLifeCycle, call mapreduce.destroy()
3. else call mapreduce.onFinish()

and I think all of that should be in a single transaction. 

Also, AbstractMapReduce.destroy() should simply call onFinish(). That means if the user extends from AbstractMapReduce and implements onFinish, it will be called through destroy. But if the user implemented the new destroy method, then he knows that onFinish is deprecated and we don't need to call it. 

It was done this way for initialize/beforeSubmit, and for consistency, it should be the same for destroy/onFinish.

**Comment by sagarkapare** [comment]
@anew The reason this is done differently than the initialize/beforeSubmit is that the onFinish method can throw an exception while the contract of the destroy method is not to throw an exception.

**Comment by sagarkapare** [comment]
@anew Updated the implementation per discussion. Please take a look. Thanks!

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT4543-1

File: cdap-api/src/main/java/co/cask/cdap/api/mapreduce/AbstractMapReduce.java [hunk 1]
```diff
@@ -177,9 +181,18 @@ protected void initialize() throws Exception {
...
+   * Classes derived from {@link AbstractMapReduce} can override this method to destroy the {@link MapReduce}.
+   * Default implementation of this method calls the deprecated {@link AbstractMapReduce#onFinish} method.
+   */
   @Override
   public void destroy() {
-    // Do nothing by default
+    try {
+      onFinish(context.getState().getStatus() == ProgramStatus.COMPLETED, context);
+    } catch (Throwable t) {
+      LOG.warn("Error executing the destroy method of the MapReduce program {}",
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/mapreduce/AbstractMapReduce.java]: it is the onFinish method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 2]
```diff
@@ -601,7 +570,11 @@ public Void call() throws Exception {
               }
             }
             context.setState(getProgramState(success, failureInfo));
-            ((ProgramLifecycle) mapReduce).destroy();
+            if (mapReduce instanceof ProgramLifecycle) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:573]: I think we need to improve the exception handling here. But that can be done as part of CDAP-4065. So it's fine for now.

**Comment by anew** [comment]
One minor comment, otherwise LGTM

**Comment by sagarkapare** [comment]
Addressed comments. New build is running here - http://builds.cask.co/browse/CDAP-DUT4543-2

**Comment by sagarkapare** [comment]
Build is green. Merging.

**Pull request merged**